### PR TITLE
fix: use commons Symbol in inlines crate

### PR DESCRIPTION
**Pull request by nfejzic**
This PR aims to refactor `unimarkup-inline` crate so that we use `Symbol` (and accompanying types) from `unimarkup-commons` crate. 

Some notes: 

* Many implemented functions on `Position`, `Span`, `Symbol` etc. were specific for inline formats. Therefore I decided to use extension traits. This might be a good idea in general. The types in `commons` crate should really have only basic implementations, everything else can be extended.
* Inline lexer and parser tests are basically non-existent at this point. There are many reasons for this, but we should probably switch to snapshot testing anyway.
* We should add few tests for position calculations. I'm not confident about correctness of that at the moment. Position is now calculated implicitly from `Symbol` positions. 
* `IntoSymbols` trait is modified for more flexibility when writing trait bound. Previously it was not possible to have bounds on output of `IntoSymbols`. Using associated type enables that. That made possible blanket implementation of `Tokenize` for `&str`, `Vec<Symbol>` and `&[Symbol]` at once.

## Edit 1: 

* Snapshot testing framework is more-or-less ready.
  * Right now, rather than extendig `test_runner` in `unimarkup_commons`, I added some 
    functions and macros to lexer tests. This seems better, as they do not generally apply
    to other stuff we might test. 
* Snapshot tests cover token recognition, token resolving and span calculations 
  beautifully!
* Markup files for lexer snapshot tests have same format as those for core tests. Idea 
  is to share these with parser snapshot tests. Using same inputs makes it easy to see
  where is the bug when it eventually happens. For example, if lexer snapshot is OK, but
  parser snapshot is not, then the bug is in the parser.
* Consequently, some functions (and/or macros) used in lexer snapshot tests could be 
  shared with parser snapshot tests. We'll decide about that during implementing parser 
  snapshot testing.

## Edit 2: 

* Snapshot testing is now implemented for inline parser as well. 
* Macros are inlined, because they were used only once. There was no need for macro then.
* Same markup files are used for inline lexer and inline parser snapshot testing.
* Setup code is shared between inline lexer and parser snapshot tests.

### CODE REVIEW

File: inline/src/lexer/token.rs [hunk 1]
```diff
@@ -294,14 +297,12 @@ impl Token {
                 any_other_kind => (any_other_kind, any_other_kind),
```
Comment (mhatzl) [inline inline/src/lexer/token.rs:297]: maybe `unreachable!()` would be better here, since the span would not be correct with the same kind twice, and the `is_ambiguous()` check should prevent this case anyways

File: inline/src/lexer/token.rs [hunk 2]
```diff
@@ -788,33 +749,47 @@ impl Span {
     ///
     /// Tuple containing the resulting span and the removed span.
     ///
-    /// [`Span`]: self::Span
+    /// [`Span`]: unimarkup_commons::scanner::span::Span
+    fn remove(self, other: Span) -> (Span, Span);
+    fn swapped(&self, other: &Span) -> (Span, Span)
```
Comment (mhatzl) [inline inline/src/lexer/token.rs:754]:
```suggestion
    fn swap(&self, other: &Span) -> (Span, Span)
```

rename as in `commons/src/scanner/span`.

File: inline/tests/lexer/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                if curr_line_nr > 0 {
+                    content.push('\n');
+                }
+
+                content.push_str(line);
+                content.push('\n');
+                curr_line_nr += 1;
+            }
+
+            {};
```
Comment (mhatzl) [inline inline/tests/lexer/snapshot.rs:29]: is this intentional?

File: inline/tests/spec/markup/plain.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  - name: breaking-unicode
+    description: |
+      Parse a character that exposed bug in some LSP clients and implementations.
+      The used character does not have the same length in UTF8 and UTF16
+
+    input: |
+      The character 𝛂 broke some stuff.
+
+    html: |
+      The character <strong>𝛂</strong> broke some stuff.
```
Comment (mhatzl) [inline inline/tests/spec/markup/plain.yml:27]: why bold?

**Review by mhatzl — CHANGES_REQUESTED** [review]
Looks really good.
I added some minor code change suggestions and remarks.

File: commons/src/scanner/mod.rs [hunk 5]
```diff
@@ -147,14 +168,48 @@ impl Symbol<'_> {
 
         &input[start..end]
     }
+
+    /// Flattens the iterator of consecutive symbols. Returns the slice of input starting from start
+    /// position of first symbol until the end of last symbol.
+    ///
+    /// Note: The input must be same in all symbols!
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:175]: I changed the wording on both `flatten` functions and API as well. API change is to return `Option` from both functions. 

The `flatten` function is fallible (for example we pass it an empty slice as param). Using `Option` as return type makes that clear to the caller and that such case must be handled.

File: commons/src/scanner/mod.rs [hunk 6]
```diff
@@ -147,14 +168,48 @@ impl Symbol<'_> {
...
+    pub fn flatten_iter<'s>(mut iter: impl Iterator<Item = &'s Symbol<'s>>) -> Option<&'s str> {
+        let first = iter.next()?;
+        let last = iter.last()?;
+
+        debug_assert!(first.input == last.input);
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:180]: I also introduced check in debug builds that all symbols are in contiguous order.

File: commons/src/scanner/mod.rs [hunk 7]
```diff
@@ -168,64 +223,59 @@ impl From<&str> for SymbolKind {
 }
 
 /// Trait for conversion of input into Unimarkup symbols.
-pub trait IntoSymbols<'s, T> {
-    /// Converts input into Unimarkup symbols.
-    fn into_symbols(self) -> T;
+pub trait IntoSymbols<'s> {
+    type Output;
+
+    /// Converts input into collection of Unimarkup symbols.
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:229]: You're right, this can return anything basically. The only restriction is at the use site, and that can guide users of the API to write the "right" implementation. 

However, I'm thinking about reworking this API completely. Aside from the problem you mentioned, here's also the problem that we have to pass it config at a later point (because of localisation). The way it's used right now won't make that possible. 

Possible solution would be to create a `Scanner` struct, with some trait (for example `trait Scan`) that creates a Scanner from input. We could even provide blanket implementation of Scanner for `T: AsRef<str>`. We can then provide functions to create create iterator from scanner, provide some collections like slice etc. That way the `Output` can be completely controlled by us, and we can have a default Config for Scanner, or pass it specific Config if needed.

What do you think?

File: commons/src/scanner/mod.rs [hunk 8]
```diff
@@ -235,20 +285,20 @@ fn word_split(input: &str) -> Vec<Symbol> {
...
+                    col_utf8: (curr_pos.col_utf8 + grapheme.len()),
+                    col_utf16: (curr_pos.col_utf16 + grapheme.encode_utf16().count()),
                     col_grapheme: (curr_pos.col_grapheme + grapheme_len),
                 }
             };
 
-            let mut kind = SymbolKind::from(word);
+            let mut kind = SymbolKind::from(grapheme);
 
             if curr_pos.col_utf8 == 1 && kind == SymbolKind::Newline {
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:296]: I don't see how length is relevant here? I'm checking if this symbols starts at the column 1 and is a `Newline`. In that case, that line is blank.

File: commons/src/scanner/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/mod.rs:229]: Could work. I guess this all depends on how we are able to handle localisation for the grapheme segmenter.

File: commons/src/scanner/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/mod.rs:296]: ah right. because col_utf8 is set to 1 after a new line.

File: commons/src/scanner/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/mod.rs:180]: have you pushed those changes already?

File: commons/src/scanner/span.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        let len_grapheme = self.len_grapheme()?;
+
+        Some(SpanLen {
+            len_utf8,
+            len_utf16,
+            len_grapheme,
+        })
+    }
+
+    pub fn swapped(&self, other: &Self) -> (Self, Self) {
```
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/span.rs:108]: `is_swapped` would seem like it should return a `bool`, but I also find it ok as `swap`. I guess one could argue that `swap` modifies the `self` in place, but then again there are other functions that behave similarly (for example [`str::replace`](https://doc.rust-lang.org/std/primitive.str.html#method.replace) creates a new string with content replaced).

TLDR: Will change it to `swap`.

[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/span.rs:108]: good point

File: commons/src/scanner/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:229]: I will try it in a separate PR, just so we can clearly see if, how and why does it improve the API.

File: commons/src/scanner/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs:180]: Not yet.

File: inline/tests/lexer/snapshot.rs (hunk 3 above)
Comment (nfejzic) [reply inline/tests/lexer/snapshot.rs:29]: No, no idea how this slipped in hahah.

File: inline/src/lexer/token.rs (hunk 2 above)
Comment (nfejzic) [reply inline/src/lexer/token.rs:754]: Oh if there's `Span` anywhere else left it must be deleted completely! Nice catch!

File: commons/src/scanner/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/mod.rs]: Pushed now.

File: inline/src/lexer/token.rs (hunk 2 above)
Comment (nfejzic) [reply inline/src/lexer/token.rs:754]: For now I will move the function to `SpanExt` trait, because we don't use that anywhere else. If necessary, we can always move it back.

File: inline/tests/spec/markup/plain.yml (hunk 4 above)
Comment (nfejzic) [reply inline/tests/spec/markup/plain.yml:27]: This was temporary, changed to plain.

File: inline/src/lexer/mod.rs [hunk 10]
```diff
@@ -549,14 +436,14 @@ impl TokenIterator<'_> {
...
         // NOTE: index points to the NEXT character, token Span is UP TO that character
-        let offset = self.index - start_pos.column;
-        let end_pos = self.pos + (0, offset);
+        let end_pos = self.get_symbol(self.index - 1)?.end;
 
-        let len = self.index - start_pos.column.saturating_sub(1);
+        let span = Span::from((start_pos, end_pos));
+        let len = span.len_grapheme()?;
 
         // TODO: improve this logic
```
Comment (mhatzl) [inline inline/src/lexer/mod.rs:444]: as internally discussed, this TODO should not be relevant anymore => TODO and commented code can be removed

**Review by mhatzl — CHANGES_REQUESTED** [review]
found one spelling mistake and a TODO that should not be relevant anymore

**Review by mhatzl — APPROVED** [review]
looks very good now 👍
